### DISCUSSION TITLE
How to use lldb with an input file?

### DISCUSSION THREAD

**Discussion by drakari7** [mfussenegger/nvim-dap#298] [Q&A]
Hello there, I want to debug code which takes inputs from an input file for c++ (or any language in general but c++ for now). Normally to execute the code I would do something like

``` cpp
./a.out < testfile
```
But how can I replicate this while being able to use breakpoints?

This debugging session does not proceed/gets stuck at `cin`.
<img width="1422" alt="image" src="https://user-images.githubusercontent.com/54357427/132959751-92354b3c-10bb-4751-8511-5a3410357269.png">

whereas this session (notice I commented out the `cin` part and gave n a value) proceeds normally
<img width="1422" alt="image" src="https://user-images.githubusercontent.com/54357427/132959812-13c2bc7d-8f5b-48bb-be51-d96452de097b.png">

**Comment by mfussenegger** [comment]
What's your configuration?

See https://github.com/mfussenegger/nvim-dap/discussions/213
Currently there isn't an option to use pipe's because the command isn't invoked within a shell.

One option would be to start the application with `gdbserver` (requires the cpptools debug adapter, see https://github.com/mfussenegger/nvim-dap/wiki/C-C---Rust-(gdb-via--vscode-cpptools)#configuration)

**Reply by drakari7** [reply]
I can't use the #213 approach because I'm not developing a proper application. I mostly wanted to use a debugger to debug competitive programming codes. I write a different file for each question, and I have to take the input with cin or scanf. My current method of debugging is printing variables and trying to figure out what's wrong. You can probably see why using a proper debugger will be much faster.

I will look into the gdbserver option, though I'm not sure how that'll work. Is there any plan for supporting piping in the future?

Also, I have not fully understood the runinterminal option in this plugin. Does this mean I can open a terminal (in neovim itself), and start the debugging session from the said terminal? Would that help in this case, as I can send the input file via the terminal perhaps?

**Reply by mfussenegger** [reply]
> Is there any plan for supporting piping in the future?

That's mostly up to debug adapters. The debug adapters are launching your application. What you define in your `dap-configurations` is forwarded to them, and they decide what to do with it.

With the exception of `runInTerminal`. In that case the debug-adapter sends back a request to nvim-dap with instructions to launch a command within a terminal. (But which command it sends, is again up to the debug-adapter, so piping is again not something nvim-dap could support by itself)

> Also, I have not fully understood the runinterminal option in this plugin. Does this mean I can open a terminal (in neovim itself), and start the debugging session from the said terminal? 

Kinda. It means if you start the debug session, it will automatically spawn a `:terminal` which runs your application.
You could also configure it to spawn an external terminal, see https://github.com/mfussenegger/nvim-dap/pull/212
